File: bench/bench_ocaml/test/test_trigo_float.ml
open Math_lib.Common
module Float = Math_lib.Float
module Trigo = Math_lib.Trigo_float

let _test_trigo_sinus ?(log = false) () =
  (* // GIVEN *)
  let error_threshold = Float.inverse (Float.new' (Z.of_int 1) (Z.of_int 12)) in
  let precision : nat = Z.of_int 11 in

  let test_sinus ((angle, expected) : Float.t * Float.t) : bool =
    let diff = Float.sub (Trigo.sinus (angle, precision)) expected in
    Float.lt diff error_threshold
  in
  (* // ASSERT *)
  let () = if log then Test.log "sin(0)" in
  let () = assert (test_sinus (Trigo.zero, Trigo.zero)) in
  let () = if log then Test.log "sin(Pi/2)" in
  let () = assert (test_sinus (Trigo.pi_half, Trigo.one)) in
  let () = if log then Test.log "sin(Pi)" in
  let () = assert (test_sinus (Trigo.pi, Trigo.zero)) in
  let () = if log then Test.log "sin(3*Pi/2)" in
  let () = assert (test_sinus (Trigo.three_pi_half, Trigo.minus_one)) in
  let () = if log then Test.log "sin(2*Pi)" in
  let () = assert (test_sinus (Trigo.two_pi, Trigo.zero)) in
  let () = if log then Test.log "sin(Pi/4)" in
  let expected = Float.div Trigo.sqrt_2 Trigo.two in
  let () = assert (test_sinus (Trigo.pi_quarter, expected)) in
  let () = if log then Test.log "sin(Pi/6)" in
  let expected = Float.inverse Trigo.two in
  let () = assert (test_sinus (Trigo.pi_sixth, expected)) in
  let () = if log then Test.log "sin(Pi/3)" in
  let expected = Float.div Trigo.sqrt_3 Trigo.two in
  let () = assert (test_sinus (Trigo.pi_third, expected)) in
  let () = if log then Test.log "sin(-Pi/2)" in
  let angle = Float.sub Trigo.zero Trigo.pi_half in
  let () = assert (test_sinus (angle, Trigo.minus_one)) in
  let () = if log then Test.log "sin(-3*Pi/2)" in
  let angle = Float.sub Trigo.pi_half Trigo.two_pi in
  let () = assert (test_sinus (angle, Trigo.one)) in

  if log then Test.log "Test 'trigo sinus (with float)' finished"

let _test_trigo_cosinus ?(log = false) () =
  (* // GIVEN *)
  let error_threshold = Float.inverse (Float.new' (Z.of_int 1) (Z.of_int 12)) in
  let precision : nat = Z.of_int 11 in

  let test_cosinus ((angle, expected) : Float.t * Float.t) : bool =
    let diff = Float.sub (Trigo.cosinus (angle, precision)) expected in
    Float.lt diff error_threshold
  in

  (* // ASSERT *)
  let () = if log then Test.log "cos(0)" in
  let () = assert (test_cosinus (Trigo.zero, Trigo.one)) in
  let () = if log then Test.log "cos(Pi/2)" in
  let () = assert (test_cosinus (Trigo.pi_half, Trigo.zero)) in
  let () = if log then Test.log "cos(Pi)" in
  let () = assert (test_cosinus (Trigo.pi, Trigo.minus_one)) in
  let () = if log then Test.log "cos(3*Pi/2)" in
  let () = assert (test_cosinus (Trigo.three_pi_half, Trigo.zero)) in
  let () = if log then Test.log "cos(2*Pi)" in
  let () = assert (test_cosinus (Trigo.two_pi, Trigo.one)) in
  let () = if log then Test.log "cos(Pi/4)" in
  let expected = Float.div Trigo.sqrt_2 Trigo.two in
  let () = assert (test_cosinus (Trigo.pi_quarter, expected)) in
  let () = if log then Test.log "cos(Pi/6)" in
  let expected = Float.div Trigo.sqrt_3 Trigo.two in
  let () = assert (test_cosinus (Trigo.pi_sixth, expected)) in
  let () = if log then Test.log "cos(Pi/3)" in
  let expected = Float.inverse (Float.new' (Z.of_int 2) (Z.of_int 0)) in
  let () = assert (test_cosinus (Trigo.pi_third, expected)) in
  let () = if log then Test.log "cos(-Pi/2)" in
  let angle = Float.sub (Float.new' (Z.of_int 0) (Z.of_int 0)) Trigo.pi_half in
  let () = assert (test_cosinus (angle, Trigo.zero)) in
  let () = if log then Test.log "cos(-3*Pi/2)" in
  let angle = Float.sub Trigo.pi_half Trigo.two_pi in
  let () = assert (test_cosinus (angle, Trigo.zero)) in

  if log then Test.log "Test 'trigo cosinus (with float)' finished"

let _test_trigo ?(log = false) () =
  let error_threshold = Float.inverse (Float.new' (Z.of_int 1) (Z.of_int 12)) in
  let precision : nat = Z.of_int 11 in
  (* // cos²(a) + sin²(a) = 1 *)
  let () = if log then Test.log "cos^2(a) + sin^2(a) = 1" in
  let angle = Trigo.pi_half in
  let expected = Float.new' (Z.of_int 1) (Z.of_int 0) in
  let cos_a = Trigo.cosinus (angle, precision) in
  let sin_a = Trigo.sinus (angle, precision) in
  let res = Float.add (Float.mul cos_a cos_a) (Float.mul sin_a sin_a) in
  let diff = Float.sub res expected in
  (* // let error = Float.resolve diff 12n in *)
  let () = assert (Float.lt diff error_threshold) in

  if log then Test.log "Test 'trigo (with float)' finished"

let all ?log () =
  _test_trigo ?log ();
  _test_trigo_sinus ?log ();
  _test_trigo_cosinus ?log ()

File: bench/bench_kotlin/src/test/kotlin/math/TestTrigoFloat.kt
package math

import math.Float.T
import math.TrigoFloat.abs
import math.TrigoFloat.add
import math.TrigoFloat.cosinus
import math.TrigoFloat.div
import math.TrigoFloat.inverse
import math.TrigoFloat.lt
import math.TrigoFloat.minus_one
import math.TrigoFloat.mul
import math.TrigoFloat.new
import math.TrigoFloat.one
import math.TrigoFloat.pi
import math.TrigoFloat.piHalf
import math.TrigoFloat.piQuarter
import math.TrigoFloat.piSixth
import math.TrigoFloat.piThird
import math.TrigoFloat.sinus
import math.TrigoFloat.sqrt2
import math.TrigoFloat.sqrt3
import math.TrigoFloat.sub
import math.TrigoFloat.threePiHalf
import math.TrigoFloat.two
import math.TrigoFloat.twoPi
import math.TrigoFloat.zero
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.State
import kotlin.test.Test
import kotlin.test.assertEquals

@State(Scope.Benchmark)
open class TestTrigoFloat {

    private val epsilon = inverse(new(1, 12))
    private val precision = 11

    private fun testSinus(angle: T, expected: T): Boolean {
        val diff = abs(sub(sinus(angle, precision))(expected))
        return lt(diff)(epsilon)
    }

    @Test
    fun testTrigoSinus() {
        assertEquals(testSinus(zero, zero), true)
        assertEquals(testSinus(piHalf, one), true)
        assertEquals(testSinus(pi, zero), true)
        assertEquals(testSinus(threePiHalf, minus_one), true)
        assertEquals(testSinus(twoPi, zero), true)
        assertEquals(testSinus(piQuarter, div(sqrt2)(two)), true)
        assertEquals(testSinus(piSixth, inverse(two)), true)
        assertEquals(testSinus(piThird, div(sqrt3)(two)), true)
        assertEquals(testSinus(sub(zero)(piHalf), minus_one), true)
        assertEquals(testSinus(sub(piHalf)(twoPi), one), true)
    }

    private fun testCosinus(angle: T, expected: T): Boolean {
        val diff = abs(sub(cosinus(angle, precision))(expected))
        return lt(diff)(epsilon)
    }

    @Test
    fun testTrigoCosinus() {
        assertEquals(testCosinus(zero, one), true)
        assertEquals(testCosinus(piHalf, zero), true)
        assertEquals(testCosinus(pi, minus_one), true)
        assertEquals(testCosinus(threePiHalf, zero), true)
        assertEquals(testCosinus(twoPi, one), true)
        assertEquals(testCosinus(piQuarter, div(sqrt2)(two)), true)
        assertEquals(testCosinus(piSixth, div(sqrt3)(two)), true)
        assertEquals(testCosinus(piThird, inverse(two)), true)
        assertEquals(testCosinus(sub(zero)(piHalf), zero), true)
        assertEquals(testCosinus(sub(piHalf)(twoPi), zero), true)
    }

    @Test
    fun testTrigo() {
        val precision = 11
        val angle = piHalf
        val cos = cosinus(angle, precision)
        val sin = sinus(angle, precision)
        val res = add(mul(cos)(cos))(mul(sin)(sin))
        val diff = abs(sub(res)(one))
        assertEquals(lt(diff)(epsilon), true);
    }

    @Benchmark
    fun all() {
        testTrigoSinus()
        testTrigoCosinus()
        testTrigo()
    }
}